src/coreclr/jit/llvmcodegen.cpp

```diff
@@ -2313,6 +2313,12 @@ void Llvm::buildIntrinsic(GenTreeIntrinsic* intrinsicNode)
     mapGenTreeToValue(intrinsicNode, intrinsicValue);
 }
 
+void Llvm::buildMemoryBarrier(GenTree* node)
+{
+    assert(node->OperIs(GT_MEMORYBARRIER));
+    _builder.CreateFence(llvm::AtomicOrdering::AcquireRelease);
```

Comment:
Just for my education, what does this actually do?  I read that it orders memory instructions?

Reply:
Yes. It is required for ordering loads/stores in the presence of multi-threaded access.

The whole picture is inordinately complicated, but the simple version is that "acquire" operations disallow moving loads/stores before them (they "acquire a lock" / "enter the critical section"), while "release" - past them (they "release the lock" / "exit the critical section"), thus "acquire + release" disallows moving memory accesses across it in either direction.